ui/src/app/routes.tsx

```diff
@@ -0,0 +1,265 @@
+  },
+  {
+    name: 'datasets',
+    path: '/workspaces/:ns/:wsid/data/datasets',
+    render: () => <WorkspaceWrapper><DataSet /></WorkspaceWrapper>,
+    title: 'Dataset Page',
+    breadcrumb: BreadcrumbType.Dataset
+  },
+  {
+    name: 'conceptSets',
+    path: '/workspaces/:ns/:wsid/concepts/sets',
+    render: () => <WorkspaceWrapper><ConceptSetList /></WorkspaceWrapper>,
+    title: 'View Concept Sets',
+    breadcrumb: BreadcrumbType.Workspace
+  },
+  {
+    name: 'conceptSet',
+    path: '/workspaces/:ns/:wsid/concepts/sets/:csid',
+    render: () => <WorkspaceWrapper><ConceptSetDetails /></WorkspaceWrapper>,
+    title: 'Concept Set',
+    breadcrumb: BreadcrumbType.ConceptSet
+  },
+  {
+    name: 'workspaceReview',
+    path: '/admin/review-workspace',
+    render: () => <AdminReviewWorkspace />,
+    title: 'Review Workspaces'
+  },
+  {
+    name: 'users',
+    path: '/admin/user',
+    render: () => <AdminUser />,
+    title: 'User Admin Table'
+  },
+  {
+    name: 'profile',
+    path: '/profile',
+    render: () => <ProfilePage />,
+    title: 'Profile'
+  },
+  {
+    name: 'settings',
+    path: '/settings',
+    render: () => <Settings />,
+    title: 'Settings'
+  },
+  {
+    name: 'workspaceCreate',
+    path: '/workspaces/build',
+    render: () => <WorkspaceEdit mode={WorkspaceEditMode.Create} />,
+    title: 'Create Workspace'
+  },
+];
+
+const handlers = routes.map(({path, ...rest}) => {
+  const keys = []; // mutated by pathToRegexp
+  const regex = pathToRegexp(path, keys);
+  return {
+    regex, ...rest,
+    keys: keys.map(k => k.name),
+    makePath: pathToRegexp.compile(path)
+  };
+});
+
+const handlersByName = fp.keyBy('name', handlers);
+
+const parseRoute = ({pathname, search}) => {
+  const handler = handlers.find(({regex}) => regex.test(pathname));
+  return handler && {
+    ...handler,
+    params: fp.zipObject(handler.keys, fp.tail(handler.regex.exec(pathname))),
+    query: qs.parse(search, {ignoreQueryPrefix: true, plainObjects: true}),
+  };
+};
+
+/**
+ * HOC that injects a `route` prop, whose value is the matched route handler, extended
+ * with `params` (path params) and `query` (query params). If no route matches, the value
+ * will be undefined.
+ */
+export const withRoute = () => WrappedComponent => {
+  class Wrapper extends React.Component<any, {location: any}> {
+    static displayName = 'withRoute()';
+    unlisten: Function;
+    constructor(props) {
+      super(props);
+      this.state = {location: history.location};
+    }
+
+    componentDidMount() {
+      this.unlisten = history.listen(location => {
+        this.setState({location});
+      });
+    }
+
+    componentWillUnmount() {
+      this.unlisten();
+    }
+
+    render() {
+      const {location} = this.state;
+      return <WrappedComponent {...this.props} route={parseRoute(location)} />;
+    }
+  }
+  return Wrapper;
+};
+
+/**
+ * Return a URL path string, given a route name and its params.
```

Comment:
This is the main function for generating urls. E.g. `getPath('notebooks', {ws, nsid})`

Navigation is done via `history.push`. We'll probably want some sort of `Link` component that abstracts this behavior nicely for links.